src/source/Crypto/Dtls_mbedtls.c

```diff
@@ -0,0 +1,739 @@
+
+    // read application data
+    while (pReadBuffer->off < pReadBuffer->len) {
+        sslRet = mbedtls_ssl_read(&pDtlsSession->sslCtx, pData + readBytes, pReadBuffer->len - pReadBuffer->off);
+        if (sslRet > 0) {
+            readBytes += sslRet;
+        } else if (sslRet == 0 || sslRet == MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY) {
+            // if sslRet is 0, the connection is closed already.
+            // if sslRet is MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY, the client notified us that the connection is going to be closed.
+            // In either case, we'll make sure that the state will change to CLOSED. If it's already closed, it'll be just a noop.
+            DLOGD("Detected DTLS close_notify alert");
+            CHK_STATUS(dtlsSessionShutdown(pDtlsSession));
+            break;
+        } else if (sslRet == MBEDTLS_ERR_SSL_WANT_READ || sslRet == MBEDTLS_ERR_SSL_WANT_WRITE) {
+            break;
+        } else {
+            LOG_MBEDTLS_ERROR("mbedtls_ssl_read", sslRet);
+            readBytes = 0;
+            retStatus = STATUS_INTERNAL_ERROR;
+            break;
+        }
+    }
+
+    if (pDtlsSession->sslCtx.state == MBEDTLS_SSL_HANDSHAKE_OVER) {
+        CHK_STATUS(dtlsSessionChangeState(pDtlsSession, CONNECTED));
+    }
+
+CleanUp:
+    if (pDataLen != NULL) {
+        *pDataLen = readBytes;
+    }
+
+    if (locked) {
+        MUTEX_UNLOCK(pDtlsSession->sslLock);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS dtlsSessionPutApplicationData(PDtlsSession pDtlsSession, PBYTE pData, INT32 dataLen)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    INT32 writtenBytes = 0;
+    BOOL locked = FALSE;
+    INT32 sslRet;
+
+    CHK(pData != NULL, STATUS_NULL_ARG);
+    CHK(!ATOMIC_LOAD_BOOL(&pDtlsSession->shutdown), retStatus);
+
+    MUTEX_LOCK(pDtlsSession->sslLock);
+    locked = TRUE;
+    
+    while (writtenBytes < dataLen) {
+        sslRet = mbedtls_ssl_write(&pDtlsSession->sslCtx, pData + writtenBytes, dataLen - writtenBytes);
+        if (sslRet > 0) {
+            writtenBytes += sslRet;
+        } else if (sslRet == MBEDTLS_ERR_SSL_WANT_READ || sslRet == MBEDTLS_ERR_SSL_WANT_WRITE) {
+            break;
+        } else {
+            LOG_MBEDTLS_ERROR("mbedtls_ssl_write", sslRet);
+            writtenBytes = 0;
+            retStatus = STATUS_INTERNAL_ERROR;
+            break;
+        }
+    }
+
+CleanUp:
+    if (locked) {
+        MUTEX_UNLOCK(pDtlsSession->sslLock);
+    }
+
+    LEAVES();
+    return STATUS_SUCCESS;
+}
+
+STATUS dtlsSessionGetLocalCertificateFingerprint(PDtlsSession pDtlsSession, PCHAR pBuff, UINT32 buffLen)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+
+    CHK(pDtlsSession != NULL && pBuff != NULL, STATUS_NULL_ARG);
+    CHK(buffLen >= CERTIFICATE_FINGERPRINT_LENGTH, STATUS_INVALID_ARG_LEN);
+
+    MUTEX_LOCK(pDtlsSession->sslLock);
+    locked = TRUE;
+
+    // TODO: Use the 0th certificate for now
```

Comment:
GitHub issue link:?

Reply:
Not really sure, we've been doing this for a while, I just copied this from the openssl implementation, https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Dtls/Dtls.c#L723-L746.